did-ion/src/sidetree.rs

```diff
@@ -0,0 +1,1926 @@
+            .context("Serialize DID Suffix Data")?;
+        Ok(Self::Long {
+            did_suffix,
+            create_operation_data: op_string,
+            _marker: PhantomData,
+        })
+    }
+}
+
+/// Convert a DID URL to an object id given a DID
+///
+/// Object id is an id of a [ServiceEndpointEntry] or [PublicKeyEntry].
+fn did_url_to_id<S: Sidetree>(did_url: &str, did: &SidetreeDID<S>) -> Result<String> {
+    let did_string = did.to_string();
+    let unprefixed = match did_url.strip_prefix(&did_string) {
+        Some(s) => s,
+        None => bail!("DID URL did not begin with expected DID"),
+    };
+    let fragment = match unprefixed.strip_prefix('#') {
+        Some(s) => s,
+        None => bail!("Expected DID URL with fragment"),
+    };
+    Ok(fragment.to_string())
+}
+
+impl<S: Sidetree> From<SidetreeDID<S>> for DIDSuffix {
+    fn from(did: SidetreeDID<S>) -> DIDSuffix {
+        match did {
+            SidetreeDID::Short { did_suffix } => did_suffix,
+            SidetreeDID::Long { did_suffix, .. } => did_suffix,
+        }
+    }
+}
+
+/// DID Resolver using ION/Sidetree REST API
+#[derive(Debug, Clone, Default)]
+pub struct HTTPSidetreeDIDResolver<S: Sidetree> {
+    pub http_did_resolver: HTTPDIDResolver,
+    pub _marker: PhantomData<S>,
+}
+
+impl<S: Sidetree> HTTPSidetreeDIDResolver<S> {
+    pub fn new(sidetree_api_url: &str) -> Self {
+        let identifiers_url = format!("{}identifiers/", sidetree_api_url);
+        Self {
+            http_did_resolver: HTTPDIDResolver::new(&identifiers_url),
+            _marker: PhantomData,
+        }
+    }
+}
+
+/// Sidetree DID Method client implementation
+pub struct SidetreeClient<S: Sidetree> {
+    pub resolver: Option<HTTPSidetreeDIDResolver<S>>,
+    pub endpoint: Option<String>,
+}
+
+impl<S: Sidetree> SidetreeClient<S> {
+    pub fn new(api_url_opt: Option<String>) -> Self {
+        let resolver_opt = api_url_opt
+            .as_ref()
+            .map(|url| HTTPSidetreeDIDResolver::new(&url));
+        Self {
+            endpoint: api_url_opt,
+            resolver: resolver_opt,
+        }
+    }
+}
+
+/// Check that a JWK is Secp256k1
+pub fn is_secp256k1(jwk: &JWK) -> bool {
+    matches!(jwk, JWK {params: ssi::jwk::Params::EC(ssi::jwk::ECParams { curve: Some(curve), ..}), ..} if curve == "secp256k1")
+}
+
+struct NoOpResolver;
+
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+impl DIDResolver for NoOpResolver {
+    async fn resolve(
+        &self,
+        _did: &str,
+        _input_metadata: &ResolutionInputMetadata,
+    ) -> (
+        ResolutionMetadata,
+        Option<Document>,
+        Option<DocumentMetadata>,
+    ) {
+        (
+            ResolutionMetadata::from_error("Missing Sidetree API endpoint"),
+            None,
+            None,
+        )
+    }
+}
+
+fn new_did_state<S: Sidetree>(
```

Comment:
This function was intended to reduce code duplication by factoring out functionality common to both the Create and Recover operations. I don't know if it would be useful to parameterize or override by Sidetree implementation.

But this function could be simplified, if the create and recover functions were changed to take one or more operations (like the update function does) instead of a verification key.